primary/src/block_synchronizer/mock.rs

```diff
@@ -0,0 +1,172 @@
+                            ready,
+                        } => {
+                            self.block_headers_expected_requests.insert(block_ids, (times, result));
+                            ready.send(()).expect("Failed to send ready message");
+                        },
+                        Core::AssertExpectations {ready} => {
+                            self.assert_expectations();
+                            ready.send(()).expect("Failed to send ready message");
+                        }
+                    }
+                }
+            }
+        }
+    }
+
+    fn assert_expectations(&self) {
+        let mut result: String = "".to_string();
+
+        for (ids, results) in &self.block_headers_expected_requests {
+            result.push_str(
+                format!(
+                    "SynchronizeBlockHeaders, ids={:?}, results={:?}",
+                    ids, results
+                )
+                .as_str(),
+            );
+        }
+
+        if !result.is_empty() {
+            panic!(
+                "There are expected calls that haven't been fulfilled \n\n {}",
+                result
+            );
+        }
+    }
+}
+
+/// A mock helper for the BlockSynchronizer to help us mock the responses
+/// eliminating the need to wire in the actual BlockSynchronizer when needed
+/// for other components.
+pub struct MockBlockSynchronizer<PublicKey: VerifyingKey> {
+    tx_core: Sender<Core<PublicKey>>,
+}
+
+impl<PublicKey: VerifyingKey> MockBlockSynchronizer<PublicKey> {
+    pub fn new(rx_commands: Receiver<Command<PublicKey>>) -> Self {
+        let (tx_core, rx_core) = channel(1);
+
+        let mut core = MockBlockSynchronizerCore {
+            block_headers_expected_requests: HashMap::new(),
+            rx_commands,
+            rx_core,
+        };
+
+        tokio::spawn(async move {
+            core.run().await;
+        });
+
+        Self { tx_core }
+    }
+
+    /// A simple method that allow us to mock responses for the
+    /// SynchronizeBlockHeaders requests.
+    /// `block_ids`: The block_ids we expect
+    /// `results`: The results we would like to respond with
+    /// `times`: How many times we should expect to be called.
+    pub async fn expect_synchronize_block_headers(
+        &self,
+        block_ids: Vec<CertificateDigest>,
+        result: Vec<BlockSynchronizeResult<BlockHeader<PublicKey>>>,
+        times: u32,
+    ) {
+        let (tx, rx) = oneshot::channel();
+        self.tx_core
+            .send(Core::SynchronizeBlockHeaders {
+                block_ids,
+                times,
+                result,
+                ready: tx,
+            })
+            .await
+            .expect("Failed to send mock expectation");
+
+        Self::await_channel(rx).await;
+    }
+
+    /// Asserts that all the expectations have been fulfilled and no
+    /// expectation has been left without having been called.
+    pub async fn assert_expectations(&self) {
```

Comment:
One of the reasons that I had to separate to `MockBlockSynchronizer` and the `MockBlockSynchronizerCore` is because I wanted to have a way to continue calling methods like this after spawning the async task (which always consumes via `move` command the object and then weren't able to call it further on tests)